core/imagedownloader.cpp

```diff
@@ -236,6 +253,17 @@ QImage Thumbnailer::fetchThumbnail(PictureEntry &entry)
 	return dummyImage;
 }
 
+void Thumbnailer::calculateThumbnails(const QVector<QString> &filenames)
+{
+	QMutexLocker l(&lock);
+	for (const QString &filename: filenames) {
+		if (!workingOn.contains(filename)) {
+			workingOn.insert(filename,
+					 QtConcurrent::run(&pool, [this, filename]() { recalculate(filename); }));
```

Comment:
For performance reasons, we don't want to calculate the same thumbnail numerous times. E.g. if the same picture was added to different dives. This is the key-point of the `workingOn` map.

The mutex simply protects the `workingOn` map. The main-thread adds, the worker threads remove entries. The threads don't wait for each other, so there should never be a chance of deadlock.

> one problem with QtConcurrent is that you can't cancel it.

Actually, you can and that is exactly what is happening here: When we add an item to be processed, we add the corresponding `QFuture` to the `workingOn` map. Currently, only one image is processed at a time, this is controlled by the
```
        pool.setMaxThreadCount(1);
```
call. If we want to cancel queued tasks, we can simply cancel the `QFuture`. This is done in `Thumbnailer::clearWorkQueue()`. This is called, if the dive is changed, so that we stop calculating thumbnails of images that are not shown. I tested it numerous times and it works like a charm.

Reply:
>> one problem with QtConcurrent is that you can't cancel it.

>  If we want to cancel queued tasks, we can simply cancel the QFuture. This is done in Thumbnailer::clearWorkQueue(). This is called, if the dive is changed, so that we stop calculating thumbnails of images that are not shown.

so i was told by Qt maintainers that a QFuture cannot be canceled if the parent class destructs before the future has finished even if `->cancel()` is called in the dtor. so this was kind of odd and i never tested it.

but in this case both Thumbnailer and ImageDownloader are singletons so i don't think this is a problem.

> The threads don't wait for each other, so there should never be a chance of deadlock.
> ...
> I tested it numerous times and it works like a charm.

ok, sounds good.

Reply:
> so i was told by Qt maintainers that a QFuture cannot be canceled if the parent class destructs before the future has finished even if ->cancel() is called in the dtor. so this was kind of odd and i never tested it.

Perhaps they mean that cancellation is asynchronous, i.e. once thumbnailing of an image is started, it will go to the end. But that is what we want - we don't want to cancel the thread during IO. `QFuture`s that are canceled before being executed are never started. At least that's what I got in testing.

As you note, the `QThreadPool` as well as the `Thumbnailer` class are long-lived, so it should pose no problem. Hopefully. :)